In [3]:
import pandas as pd
pd.__version__

'2.0.2'

***Работа с различными источниками данных в Pandas***

In [4]:
countries_df = pd.DataFrame({
    'country': ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан'],
    'population': [56.29, 38.05, 322.28, 146.24, 45.5, 9.5, 17.04],
    'square': [133396, 9984670, 9826630, 17125191, 603628, 207600, 2724902]
})

countries_df.to_csv('data/countries.csv', index=False, sep=';')

Убедимся, что сохранённый нами ранее файл создался верно. Для этого прочитаем его в переменную **countries_data** и выведем её на экран. Не забудем также о том, что мы использовали в качестве разделителя **';'**:

In [5]:
countries_data = pd.read_csv('data/countries.csv', sep=';')
display(countries_data)

country  population    square
0     Англия       56.29    133396
1     Канада       38.05   9984670
2        США      322.28   9826630
3     Россия      146.24  17125191
4    Украина       45.50    603628
5   Беларусь        9.50    207600
6  Казахстан       17.04   2724902

**ЧТЕНИЕ CSV-ФАЙЛА ПО ССЫЛКЕ**

In [6]:
data = pd.read_csv('https://raw.githubusercontent.com/esabunor/MLWorkspace/master/melb_data.csv')
display(data)

Unnamed: 0        Suburb           Address  Rooms Type      Price  \
0               1    Abbotsford      85 Turner St      2    h  1480000.0   
1               2    Abbotsford   25 Bloomburg St      2    h  1035000.0   
2               4    Abbotsford      5 Charles St      3    h  1465000.0   
3               5    Abbotsford  40 Federation La      3    h   850000.0   
4               6    Abbotsford       55a Park St      4    h  1600000.0   
...           ...           ...               ...    ...  ...        ...   
18391       23540  Williamstown   8/2 Thompson St      2    t   622500.0   
18392       23541  Williamstown      96 Verdon St      4    h  2500000.0   
18393       23544     Yallambie      17 Amaroo Wy      4    h  1100000.0   
18394       23545    Yarraville        6 Agnes St      4    h  1285000.0   
18395       23546    Yarraville     33 Freeman St      4    h  1050000.0   

      Method     SellerG        Date  Distance  ...  Bathroom  Car  Landsize  \
0          S      Biggin   3/12/2016       2.5  ...       1.0  1.0     202.0   
1          S      Biggin   4/02/2016       2.5  ...       1.0  0.0     156.0   
2         SP      Biggin   4/03/2017       2.5  ...       2.0  0.0     134.0   
3         PI      Biggin   4/03/2017       2.5  ...       2.0  1.0      94.0   
4         VB      Nelson   4/06/2016       2.5  ...       1.0  2.0     120.0   
...      ...         ...         ...       ...  ...       ...  ...       ...   
18391     SP        Greg  26/08/2017       6.8  ...       2.0  1.0       NaN   
18392     PI     Sweeney  26/08/2017       6.8  ...       1.0  5.0     866.0   
18393      S  Buckingham  26/08/2017      12.7  ...       3.0  2.0       NaN   
18394     SP     Village  26/08/2017       6.3  ...       1.0  1.0     362.0   
18395     VB     Village  26/08/2017       6.3  ...       2.0  2.0       NaN   

       BuildingArea  YearBuilt  CouncilArea  Lattitude Longtitude  \
0               NaN        NaN        Yarra  -37.79960  144.99840   
1              79.0     1900.0        Yarra  -37.80790  144.99340   
2             150.0     1900.0        Yarra  -37.80930  144.99440   
3               NaN        NaN        Yarra  -37.79690  144.99690   
4             142.0     2014.0        Yarra  -37.80720  144.99410   
...             ...        ...          ...        ...        ...   
18391          89.0     2010.0          NaN  -37.86393  144.90484   
18392         157.0     1920.0          NaN  -37.85908  144.89299   
18393           NaN        NaN          NaN  -37.72006  145.10547   
18394         112.0     1920.0          NaN  -37.81188  144.88449   
18395         139.0     1950.0          NaN  -37.81829  144.87404   

                  Regionname  Propertycount  
0      Northern Metropolitan         4019.0  
1      Northern Metropolitan         4019.0  
2      Northern Metropolitan         4019.0  
3      Northern Metropolitan         4019.0  
4      Northern Metropolitan         4019.0  
...                      ...            ...  
18391   Western Metropolitan         6380.0  
18392   Western Metropolitan         6380.0  
18393  Northern Metropolitan         1369.0  
18394   Western Metropolitan         6543.0  
18395   Western Metropolitan         6543.0  

[18396 rows x 22 columns]

*Прочитаем наши данные о недвижимости из **csv-файла** и запишем результирующий DataFrame в переменную **melb_data**:*

In [31]:
import pandas as pd
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
display(melb_data)

index         Suburb           Address  Rooms Type      Price Method  \
0          0     Abbotsford      85 Turner St      2    h  1480000.0      S   
1          1     Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2          2     Abbotsford      5 Charles St      3    h  1465000.0     SP   
3          3     Abbotsford  40 Federation La      3    h   850000.0     PI   
4          4     Abbotsford       55a Park St      4    h  1600000.0     VB   
...      ...            ...               ...    ...  ...        ...    ...   
13575  13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
0        Biggin   3/12/2016       2.5  ...  1.0     202.0         126.0   
1        Biggin   4/02/2016       2.5  ...  0.0     156.0          79.0   
2        Biggin   4/03/2017       2.5  ...  0.0     134.0         150.0   
3        Biggin   4/03/2017       2.5  ...  1.0      94.0         126.0   
4        Nelson   4/06/2016       2.5  ...  2.0     120.0         142.0   
...         ...         ...       ...  ...  ...       ...           ...   
13575     Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576  Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577     Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578   Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579   Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
0         1970.0        Yarra  -37.79960   144.99840   
1         1900.0        Yarra  -37.80790   144.99340   
2         1900.0        Yarra  -37.80930   144.99440   
3         1970.0        Yarra  -37.79690   144.99690   
4         2014.0        Yarra  -37.80720   144.99410   
...          ...          ...        ...         ...   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
0           Northern Metropolitan         4019.0    -37.7996, 144.9984  
1           Northern Metropolitan         4019.0    -37.8079, 144.9934  
2           Northern Metropolitan         4019.0    -37.8093, 144.9944  
3           Northern Metropolitan         4019.0    -37.7969, 144.9969  
4           Northern Metropolitan         4019.0    -37.8072, 144.9941  
...                           ...            ...                   ...  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[13580 rows x 23 columns]

***Задание 5.1***

*Какова цена объекта недвижимости под индексом 15?*

In [34]:
import pandas as pd
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
melb_data.loc[15, 'Price']

1310000.0

***Задание 5.2***

*Когда был продан объект под индексом 90?*

In [35]:
import pandas as pd
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
melb_data.loc[90, 'Date']

'10/09/2016'

***Задание 5.3***

*Во сколько раз площадь прилегающей территории, на которой находится здание с индексом 3521, больше площади участка, на котором находится здание с индексом 1690? Ответ округлите до целого числа.*

In [37]:
import pandas as pd
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
round(melb_data.loc[3521, 'Landsize'] / melb_data.loc[1690, 'Landsize'])

3

**Исследование структуры DataFrame**

***Задание 6.1***

Дана таблица **data = pd.DataFrame([[0,1], [1, 0], [1, 1]]**, **columns=['А', 'B'])**. Что выведет код **print(data.shape)**?

In [39]:
data = pd.DataFrame([[0,1], [1, 0], [1, 1]], columns=['А', 'B'])
print(data.shape)

(3, 2)


***Задание 6.2***

Какой из нижеперечисленных вариантов кода позволяет получить последние 15 строк DataFrame melb_data?

In [40]:
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
melb_data.tail(15)

index          Suburb             Address  Rooms Type      Price  \
13565  13565         Vermont      73 Nurlendi Rd      4    h  1263500.0   
13566  13566   Vermont South       9 Winswood Cl      4    h  1250000.0   
13567  13567        Viewbank       149 Graham Rd      5    h  1316000.0   
13568  13568        Wantirna  16 chesterfield Ct      4    h   951000.0   
13569  13569  Wantirna South        12 Armagh Cr      4    h  1323000.0   
13570  13570  Wantirna South       34 Fewster Dr      3    h   970000.0   
13571  13571  Wantirna South          15 Mara Cl      4    h  1330000.0   
13572  13572        Watsonia       76 Kenmare St      2    h   650000.0   
13573  13573        Werribee         5 Nuragi Ct      4    h   635000.0   
13574  13574     Westmeadows          9 Black St      3    h   582000.0   
13575  13575   Wheelers Hill        12 Strada Cr      4    h  1245000.0   
13576  13576    Williamstown       77 Merrett Dr      3    h  1031000.0   
13577  13577    Williamstown         83 Power St      3    h  1170000.0   
13578  13578    Williamstown        96 Verdon St      4    h  2500000.0   
13579  13579      Yarraville          6 Agnes St      4    h  1285000.0   

      Method        SellerG        Date  Distance  ...  Car  Landsize  \
13565      S          Barry  26/08/2017      17.2  ...  2.0     791.0   
13566     PI         Philip  26/08/2017      17.2  ...  2.0     986.0   
13567     SP         Nelson  26/08/2017       8.9  ...  3.0     696.0   
13568      S            Ray  26/08/2017      14.7  ...  2.0     704.0   
13569      S          Barry  26/08/2017      14.7  ...  2.0     551.0   
13570      S          Barry  26/08/2017      14.7  ...  2.0     674.0   
13571      S          Barry  26/08/2017      14.7  ...  2.0     717.0   
13572     PI       Morrison  26/08/2017      14.5  ...  1.0     210.0   
13573      S  hockingstuart  26/08/2017      14.7  ...  1.0     662.0   
13574      S            Red  26/08/2017      16.5  ...  2.0     256.0   
13575      S          Barry  26/08/2017      16.7  ...  2.0     652.0   
13576     SP       Williams  26/08/2017       6.8  ...  2.0     333.0   
13577      S          Raine  26/08/2017       6.8  ...  4.0     436.0   
13578     PI        Sweeney  26/08/2017       6.8  ...  5.0     866.0   
13579     SP        Village  26/08/2017       6.3  ...  1.0     362.0   

       BuildingArea  YearBuilt  CouncilArea  Lattitude  Longtitude  \
13565         126.0     1970.0          NaN  -37.84467   145.19276   
13566         126.0     1970.0          NaN  -37.84679   145.20051   
13567         126.0     1970.0          NaN  -37.73501   145.08341   
13568         200.0     1981.0          NaN  -37.85639   145.23536   
13569         126.0     1970.0          NaN  -37.87608   145.22390   
13570         126.0     1970.0          NaN  -37.88360   145.22805   
13571         191.0     1980.0          NaN  -37.86887   145.22116   
13572          79.0     2006.0          NaN  -37.70657   145.07878   
13573         172.0     1980.0          NaN  -37.89327   144.64789   
13574         126.0     1970.0          NaN  -37.67917   144.89390   
13575         126.0     1981.0          NaN  -37.90562   145.16761   
13576         133.0     1995.0          NaN  -37.85927   144.87904   
13577         126.0     1997.0          NaN  -37.85274   144.88738   
13578         157.0     1920.0          NaN  -37.85908   144.89299   
13579         112.0     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
13565        Eastern Metropolitan         4181.0  -37.84467, 145.19276  
13566        Eastern Metropolitan         4280.0  -37.84679, 145.20051  
13567        Eastern Metropolitan         2698.0  -37.73501, 145.08341  
13568        Eastern Metropolitan         5424.0  -37.85639, 145.23536  
13569        Eastern Metropolitan         7082.0   -37.87608, 145.2239  
13570        Eastern Metropolitan         7082.0   -37.8836, 145.22805  
13571

**Правильный ответ:** melb_data.tail(15)

***Задание 6.3***

Какой тип данных имеет столбец Coordinates?

In [48]:
melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  float64
 12  Bathroom       13580 non-null  float64
 13  Car            13580 non-null  float64
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  float64
 17  CouncilArea    12211 non-null  object 
 18  Lattit

**Правильный ответ:** object

***Задание 6.4***

Какое утверждение о типе данных **object** верно?

**Правильный ответ:** Это тип данных, которым в Pandas кодируются текстовые или другие нечисловые значения (кортежи, списки и т. д.)

***Задание 6.5***

*Сколько пропущенных значений в столбце ***CouncilArea***?*

***Правильный ответ: 1369***

***Задание 6.6***

*Сколько столбцов после преобразования типов имеет целочисленный тип данных?*

***Правильный ответ: 8***

***Задание 6.7***

Выберите верные утверждения в отношении таблицы **melb_data**:

**Правильные ответы:**

**1)** Половина проданных зданий находится на расстоянии от 6.1 до 13 километров от центра города (Квартиль уровня 0.25 столбца Distance равен 6.1 км, квартиль уровня 0.75 того же столбца равен 13 км.)

**2)** В данных есть как минимум одно жилое помещение, которое находится ровно в той точке, которая считается центром Мельбурна (Минимальное расстояние (столбец Distance) от центра города равно 0. Это значит, что на карте есть как минимум одна точка, которая совпадает с точкой, обозначенной как центр города.)

***Задание 6.8***

Выберите верные утверждения в отношении таблицы **melb_data**:

**Правильные ответы:**

**1)** Мельбурн и его пригороды разделены как минимум на восемь районов (В столбце Regionname восемь уникальных значений (unique), что даёт право сделать вывод, что в Мельбурне и его пригородах как минимум восемь различных районов.)

**2)** У некоторых различных адресов в таблице совпадают географические координаты (Количество уникальных значений (unique) в столбце Address больше, чем количество уникальных значений в столбце Coordinates. Это значит, что некоторые различные адреса помечены одинаковыми географическими координатами.)

**3)** В пригороде Reservoir расположено наибольшее число проданных объектов - 359 (Наиболее часто встречающееся (top) значение в столбце Suburb является Reservoir. Частота (freq) его использований - 359 раз.)


***Задание 6.9***

Сколько процентов от общего количества домов составляют таунхаусы (тип объекта - t)?

*Ответ округлите до целого числа. Не указывайте знак %.*

In [58]:
melb_data['Type'].value_counts(normalize=True)

Type
h    0.695803
u    0.222165
t    0.082032
Name: proportion, dtype: float64

**Правильный ответ:** 8

**Статистические методы**

***Задание 7.2***

Чему равно максимальное количество домов на продажу в районе (Propertycount)?

In [1]:
import pandas as pd
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
display(melb_data)

index         Suburb           Address  Rooms Type      Price Method  \
0          0     Abbotsford      85 Turner St      2    h  1480000.0      S   
1          1     Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2          2     Abbotsford      5 Charles St      3    h  1465000.0     SP   
3          3     Abbotsford  40 Federation La      3    h   850000.0     PI   
4          4     Abbotsford       55a Park St      4    h  1600000.0     VB   
...      ...            ...               ...    ...  ...        ...    ...   
13575  13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
0        Biggin   3/12/2016       2.5  ...  1.0     202.0         126.0   
1        Biggin   4/02/2016       2.5  ...  0.0     156.0          79.0   
2        Biggin   4/03/2017       2.5  ...  0.0     134.0         150.0   
3        Biggin   4/03/2017       2.5  ...  1.0      94.0         126.0   
4        Nelson   4/06/2016       2.5  ...  2.0     120.0         142.0   
...         ...         ...       ...  ...  ...       ...           ...   
13575     Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576  Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577     Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578   Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579   Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
0         1970.0        Yarra  -37.79960   144.99840   
1         1900.0        Yarra  -37.80790   144.99340   
2         1900.0        Yarra  -37.80930   144.99440   
3         1970.0        Yarra  -37.79690   144.99690   
4         2014.0        Yarra  -37.80720   144.99410   
...          ...          ...        ...         ...   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
0           Northern Metropolitan         4019.0    -37.7996, 144.9984  
1           Northern Metropolitan         4019.0    -37.8079, 144.9934  
2           Northern Metropolitan         4019.0    -37.8093, 144.9944  
3           Northern Metropolitan         4019.0    -37.7969, 144.9969  
4           Northern Metropolitan         4019.0    -37.8072, 144.9941  
...                           ...            ...                   ...  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[13580 rows x 23 columns]

In [2]:
melb_data['Propertycount'].max()

21650.0

***Задание 7.3***

Чему равно стандартное отклонение (разброс) расстояния от центра города до объекта недвижимости?

In [3]:
round(melb_data['Distance'].std())

6

***Задание 7.4***

Чему равно отклонение (в процентах) медианного значения площади здания от его среднего значения?

In [4]:
building_median = melb_data['BuildingArea'].median() 
building_mean =  melb_data['BuildingArea'].mean()
deviance = abs(building_median - building_mean)/building_mean
print(round(deviance * 100, 2))

9.76


***Задание 7.5***

Задан ряд чисел [1, 2, 4, 2, 3, 2, 1, 5, 6]. Чему равна мода в данном ряду?

**Правильный ответ:** 2

***Задание 7.6***

Сколько спален чаще всего встречается в домах в Мельбурне?

In [8]:
melb_data['Bedroom'].mode()

0    3.0
Name: Bedroom, dtype: float64

**Фильтрация данных в DataFrame**

***Задание 8.1***

У скольких объектов недвижимости из таблицы melb_data отсутствуют ванные комнаты?

In [9]:
melb_data[melb_data['Bathroom'] == 0].shape[0]

34

***Задание 8.2***

Сколько в таблице melb_data объектов недвижимости, которые были проданы риелторской компанией Nelson и стоимость которых составила больше 3 миллионов?

In [10]:
melb_data[(melb_data['SellerG'] == 'Nelson') & (melb_data['Price'] > 3e6)].shape[0]

5

***Задание 8.3***

Какова минимальная стоимость участка без здания (площадь здания равна 0) в таблице melb_data?

In [11]:
melb_data[(melb_data['BuildingArea'] == 0)]['Price'].min()

412500.0

***Задание 8.4***

Какова средняя цена объектов недвижимости в таблице melb_data с ценой менее одного миллиона, в которых либо количество комнат больше пяти, либо здание моложе 2015 года?

In [12]:
round(melb_data[(melb_data['Price']<1e6) & ((melb_data['Rooms']>5) | (melb_data['YearBuilt'] > 2015))]['Price'].mean())

769239

***Задание 8.5***

В каком районе Мельбурна чаще всего продаются виллы и коттеджи (тип здания — h) с ценой меньше трёх миллионов?

In [13]:
melb_data[(melb_data['Type'] == 'h') & (melb_data['Price'] < 3000000)]['Regionname'].mode()

0    Northern Metropolitan
Name: Regionname, dtype: object

**Закрепление знаний**

***ФИНАЛЬНОЕ ЗАДАНИЕ:***

In [62]:
import pandas as pd
pd.__version__

'2.0.2'

In [65]:
student_data = pd.read_csv('data/students_performance.csv', sep=',')
display(student_data)

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
0                      none          72             72             74  
1                 completed          69             90             88  
2                      none          90             95             93  
3                      none          47             57             44  
4                      none          76             78             75  
..                      ...         ...            ...            ...  
995               completed          88             99             95  
996                    none          62             55             55  
997               completed          59             71             65  
998               completed          68             78             77  
999                    none          77             86             86  

[1000 rows x 8 columns]

***Задание 9.1***

Данные о скольких студентах содержатся в таблице?

In [68]:
student_data.shape[0]

1000

***Задание 9.2***

Каков балл по письму у студента под индексом 155?

In [67]:
student_data.loc[155, 'writing score']

88

***Задание 9.3***

Сколько суммарно пропущенных значений в таблице?

In [69]:
student_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


***Задание 9.4***

Сколько столбцов в таблице имеет тип данных object?

In [70]:
student_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


***Задание 9.5***

Какой объём памяти (в килобайтах) занимает таблица?

In [71]:
student_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


***Задание 9.6***

Каков у студентов средний балл по математике?

In [80]:
student_data['math score'].mean()

66.089

***Задание 9.7***

Какая расовая группа является самой крупной в учебном заведении?

In [81]:
student_data['race/ethnicity'].mode()

0    group C
Name: race/ethnicity, dtype: object

***Задание 9.8***

Каков средний балл по чтению у студентов, которые посещали курсы подготовки к экзаменам?

In [75]:
round(student_data[student_data['test preparation course'] == 'completed']['reading score'].mean())

74

***Задание 9.9***

Сколько студентов получили 0 баллов по математике?

In [76]:
student_data[student_data['math score'] == 0].shape[0]

1

***Задание 9.10***

Проверьте гипотезу: у студентов с оплачиваемым питанием средний балл по математике выше, чем у студентов с льготным питанием.*
*В качестве ответа напишите наибольший средний балл по математике среди этих групп студентов.

In [83]:
student_data[student_data['lunch'] == 'standard']['math score'].mean()

70.03410852713178

In [79]:
student_data[student_data['lunch'] == 'free/reduced']['math score'].mean()

58.92112676056338

***Задание 9.11***

Каков процент студентов, родители которых имеют высшее образование уровня бакалавриата (bachelor's degree)?

In [84]:
student_data["parental level of education"].value_counts(normalize=True)

parental level of education
some college          0.226
associate's degree    0.222
high school           0.196
some high school      0.179
bachelor's degree     0.118
master's degree       0.059
Name: proportion, dtype: float64

***Задание 9.12***

Насколько медианный балл по письму у студентов в расовой группе А отличается от среднего балла по письму у студентов в расовой группе C?

In [85]:
a = student_data[student_data['race/ethnicity'] == 'group A']['writing score'].median()
b = student_data[student_data['race/ethnicity'] == 'group C']['writing score'].mean()
print(round(abs(a - b)))

6
